mypy/meet.py

```diff
@@ -54,6 +54,12 @@ def narrow_declared_type(declared: Type, narrowed: Type) -> Type:
         return TypeType.make_normalized(narrow_declared_type(declared.item, narrowed.item))
     elif isinstance(declared, (Instance, TupleType, TypeType, LiteralType)):
         return meet_types(declared, narrowed)
+    elif isinstance(declared, TypedDictType) and isinstance(narrowed, Instance):
+        # Special case useful for selecting TypedDicts from unions using isinstance(x, dict).
+        if (narrowed.type.fullname() == 'builtins.dict' and
+                all(isinstance(t, AnyType) for t in narrowed.args)):
+            return declared
+        return narrowed
```

Comment:
Hmm, I'm not sure this is the best thing to do, but it's consistent with what we did previously. One option would be to call recursively and narrow the fallback type instead. It's okay to leave like this for now, but maybe at least add a TODO comment.

Reply:
I can try adding the recursive call unless this will cause troubles in internal code bases.